examples/ethereum-tracker/src/contract.rs

```diff
@@ -69,69 +75,96 @@ impl Contract for EthereumTrackerContract {
-                start_block,
-                start_block + 1,
-            )
-            .await
-            .expect("Read the Initial event");
-        assert_eq!(events.len(), 1);
-        let event = events[0].clone();
-        let EthereumDataType::Address(address) = event.values[0].clone() else {
-            panic!("wrong type for the first entry");
+        let application_id = self.runtime.application_id();
+        let response = self.runtime.query_service(application_id, request);
+
+        let async_graphql::Value::Object(data_object) = response.data else {
+            panic!("Unexpected response from `readInitialEvent`: {response:#?}");
         };
-        let EthereumDataType::Uint256(value) = event.values[1] else {
-            panic!("wrong type for the second entry");
+        let async_graphql::Value::Object(ref initial_event) = data_object["readInitialEvent"]
+        else {
+            panic!("Unexpected response data from `readInitialEvent`: {data_object:#?}");
         };
-        let value = U256Cont { value };
-        self.state.accounts.insert(&address, value).unwrap();
+        let async_graphql::Value::String(ref address) = initial_event["address"] else {
+            panic!("Unexpected address in initial event: {initial_event:#?}");
+        };
+        let async_graphql::Value::String(ref balance_string) = initial_event["balance"] else {
+            panic!("Unexpected balance in initial event: {initial_event:#?}");
+        };
+
+        assert!(
+            balance_string.starts_with("0x"),
+            "Balance is not a hex string starting with `0x`"
+        );
+
+        let balance = U256::from_str_radix(&balance_string[2..], 16)
+            .expect("Balance could not be parsed as a hex string");
+
+        self.state
+            .accounts
+            .insert(address, balance.into())
+            .expect("Failed to insert initial balance");
     }
 
-    async fn update(&mut self, to_block: u64) {
-        let event_name_expanded = "Transfer(address indexed,address indexed,uint256)";
-        let (ethereum_client, contract_address) = self.get_endpoints();
-        let start_block = self.state.start_block.get_mut();
-        let events = ethereum_client
-            .read_events(
-                &contract_address,
-                event_name_expanded,
-                *start_block,
-                to_block,
-            )
-            .await
-            .expect("Read a transfer event");
-        *start_block = to_block;
-        for event in events {
-            let EthereumDataType::Address(from) = event.values[0].clone() else {
-                panic!("wrong type for the first entry");
+    /// Updates the accounts based on the transfer events emitted up to the `end_block`.
+    async fn update(&mut self, end_block: u64) {
+        let request = async_graphql::Request::new(format!(
+            r#"query {{ readTransferEvents(endBlock: {end_block}) }}"#
+        ));
```

Comment:
The code in main is reading events from a `start_block` to a `to_block` (renamed `end_block` here).
In the new code, there isn't anymore a `start_code`. How do we avoid repeating the operations?

Reply:
The service has access to the state, too, and `read_transfer_events` uses the `start_block` field. No need to pass it in here.